In [1]:
import numpy as np
import code_for_hw5 as hw5
import auto as oto

### Datasets and Polynomial features transformation

In [2]:
X_s1 = hw5.make_polynomial_feature_fun(1)(oto.auto_data[0])
X_s2 = hw5.make_polynomial_feature_fun(1)(oto.auto_data[1])

In [3]:
Y_s1 = oto.auto_values
Y_s2 = oto.auto_values

In [4]:
def test(X, Y, lams, k=10):
    for lam in lams:
       score = hw5.xval_learning_alg(X, Y, lam, k).tolist() 
       print(" For lam = {}, score = {} ".format(lam, score))
    

### Test cases on ridge_min for order 1 and set 1

In [5]:
lams = np.arange(0.0, 0.11, 0.01).tolist()
test(X_s1, Y_s1, lams)

 For lam = 0.0, score = [[0.5482321779790428]] 
 For lam = 0.01, score = [[0.5483481494451308]] 
 For lam = 0.02, score = [[0.5484706315938134]] 
 For lam = 0.03, score = [[0.5485995147151684]] 
 For lam = 0.04, score = [[0.5487346905794944]] 
 For lam = 0.05, score = [[0.5488760524229521]] 
 For lam = 0.06, score = [[0.5490234949331569]] 
 For lam = 0.07, score = [[0.5491769142347266]] 
 For lam = 0.08, score = [[0.5493362078747925]] 
 For lam = 0.09, score = [[0.5495012748084831]] 
 For lam = 0.1, score = [[0.5496720153843853]] 


### Test cases on ridge_min for order 1 and set 2

In [6]:
test(X_s2, Y_s2, lams)

 For lam = 0.0, score = [[0.5313179818108111]] 
 For lam = 0.01, score = [[0.5315133134098202]] 
 For lam = 0.02, score = [[0.531717019784605]] 
 For lam = 0.03, score = [[0.5319289402037708]] 
 For lam = 0.04, score = [[0.5321489162382212]] 
 For lam = 0.05, score = [[0.5323767917423792]] 
 For lam = 0.06, score = [[0.532612412835018]] 
 For lam = 0.07, score = [[0.5328556278797321]] 
 For lam = 0.08, score = [[0.5331062874650659]] 
 For lam = 0.09, score = [[0.5333642443843273]] 
 For lam = 0.1, score = [[0.5336293536151068]] 


### Test cases on ridge_min for order 2 and s1

In [7]:
X_s1 = hw5.make_polynomial_feature_fun(2)(oto.auto_data[0])
X_s2 = hw5.make_polynomial_feature_fun(2)(oto.auto_data[1])

In [8]:
test(X_s1, Y_s1, lams)

 For lam = 0.0, score = [[0.5165256640552364]] 
 For lam = 0.01, score = [[0.5166289755089962]] 
 For lam = 0.02, score = [[0.5167397716973583]] 
 For lam = 0.03, score = [[0.5168579146246236]] 
 For lam = 0.04, score = [[0.516983268281214]] 
 For lam = 0.05, score = [[0.5171156986251995]] 
 For lam = 0.06, score = [[0.5172550735636575]] 
 For lam = 0.07, score = [[0.5174012629338752]] 
 For lam = 0.08, score = [[0.5175541384844106]] 
 For lam = 0.09, score = [[0.5177135738560216]] 
 For lam = 0.1, score = [[0.5178794445624765]] 


### Test cases on ridge_min for order 2 and s2

In [9]:
test(X_s2, Y_s2, lams)

 For lam = 0.0, score = [[0.4983126421670579]] 
 For lam = 0.01, score = [[0.49840557061366153]] 
 For lam = 0.02, score = [[0.49850492687879794]] 
 For lam = 0.03, score = [[0.49861059531193563]] 
 For lam = 0.04, score = [[0.4987224619815008]] 
 For lam = 0.05, score = [[0.4988404146552872]] 
 For lam = 0.06, score = [[0.49896434278088825]] 
 For lam = 0.07, score = [[0.49909413746616327]] 
 For lam = 0.08, score = [[0.4992296914597424]] 
 For lam = 0.09, score = [[0.4993708991315783]] 
 For lam = 0.1, score = [[0.4995176564535476]] 


### Test cases on ridge_min for order 3 and set 1

In [10]:
X_s1 = hw5.make_polynomial_feature_fun(3)(oto.auto_data[0])
X_s2 = hw5.make_polynomial_feature_fun(3)(oto.auto_data[1])

In [11]:
lams = np.arange(0, 220, 20).tolist()
test(X_s1, Y_s1, lams)

 For lam = 0, score = [[11694443.349454524]] 
 For lam = 20, score = [[0.8311179838598667]] 
 For lam = 40, score = [[0.7728220013230147]] 
 For lam = 60, score = [[0.7747701080121777]] 
 For lam = 80, score = [[0.773896926936408]] 
 For lam = 100, score = [[0.773731032442597]] 
 For lam = 120, score = [[0.7757175165287513]] 
 For lam = 140, score = [[0.7798219587144657]] 
 For lam = 160, score = [[0.7860549559829583]] 
 For lam = 180, score = [[0.7951042565623607]] 
 For lam = 200, score = [[0.8051893857296257]] 


### Test cases on ridge_min for order 3 and set 2

In [12]:
test(X_s2, Y_s2, lams)

 For lam = 0, score = [[479995.8342373061]] 
 For lam = 20, score = [[0.7359266197078987]] 
 For lam = 40, score = [[0.7583502352905847]] 
 For lam = 60, score = [[0.7696464214712557]] 
 For lam = 80, score = [[0.7757159150012228]] 
 For lam = 100, score = [[0.7811950432479638]] 
 For lam = 120, score = [[0.7873332989226618]] 
 For lam = 140, score = [[0.7944650405515509]] 
 For lam = 160, score = [[0.8026517054031356]] 
 For lam = 180, score = [[0.8119692191645969]] 
 For lam = 200, score = [[0.8225565007185475]] 


### `Conclusion`
##### The combinaison that minimizes the RMSE is (2, 2, 0.0) / (set 2, order 2, lambda). Lambda has less impact on RMSE.

### Cross-validation RMSE value (in mpg) that you obtain using the best combination

In [13]:
X_s2 = hw5.make_polynomial_feature_fun(2)(oto.auto_data[1])
rmse = hw5.xval_learning_alg(X_s2, Y_s2, 0.0, 10)
rmse_mpg = rmse * oto.sigma
rmse_mpg


array([[3.88436985]])

### Best lambda and corresponding RMSE_mpg by fiting an 3 order polynomial model using the first set feature.


#### From previous tests, we find that for lambda =  `40`, the  rmse is minimum ansd rmse = `0.7728220013230147`. So the rmse in mpg is 

In [14]:
rmse = hw5.xval_learning_alg(X_s1, Y_s1, 40, 10)
rmse_mpg = rmse * oto.sigma
rmse_mpg

array([[6.02418287]])

In [15]:
rmse = hw5.xval_learning_alg(X_s1, Y_s1, 26, 10)
rmse_mpg = rmse * oto.sigma
rmse_mpg

array([[5.88292896]])